`.resample`: faster convolution by sparse matrices

aitap opened the pull request:
The nested for-loop I had suggested back in 2020 is very slow:

```r
library(txtplot)
N <- 20
system.time(for (i in 1:N) txtimage(matrix(10, 200, 200))) / N
# current version:
#    user  system elapsed
# 0.58680 0.00605 0.59285
```

We can speed it up at the cost of introducing a dependency on the recommended package [Matrix](https://cran.r-project.org/package=Matrix), which is admittedly slow to load by itself:

```r
system.time(library(txtplot))
# new version:
#    user  system elapsed
#   0.644   0.039   0.684
N <- 20
system.time(for (i in 1:N) txtimage(matrix(10, 200, 200))) / N
#    user  system elapsed
# 0.05915 0.01305 0.07220
```

What used to be a nested for loop (for every row `i`, every column `j`, compute the Lanczos kernels, sum the elementwise product, normalise) can be sped up by using more matrix operations:

1. Compute per-row and per-column Lanczos kernels, store them in sparse matrices
2. Use a Kronecker product to obtain a giant $O(\mathtt{length}(\mathtt{z})\cdot{}\mathtt{width}\cdot{}\mathtt{height})$ but sparse matrix containing every kernel needed to compute the resampled matrix
3. Use a matrix product to resample the matrix in one operation
4. Propagate the missingness from the source matrix

What do you think? I really appreciate packages that have no non-base dependencies, and while Matrix has `Priority: recommended` and should therefore be installed almost everywhere where R is, the dependency count would no longer be zero. If you have other ideas regarding signal-conservative image resizing in base R, please let me know.

I can also delay loading `Matrix` by using calls of the form `Matrix::Matrix` instead of `importFrom(Matrix, ...)` in the `NAMESPACE`. This would help `txtplot` load as fast as it currently does, but the first call to `txtimage(...)` would be delayed by half a second.

aitap committed:
.resample: convolution by sparse matrices

Instead of spending time in a nested for loop, put the Lancoz kernels
into sparse matrices, compute the full kernel matrix using a Kronecker
product, then multiply by the resulting matrix to obtain the desired
resampled matrix.

bbnkmp commented:
Great, thanks a lot for this improvement!
Usually I would not worry about package start-up time, but for this package it may matter.
Can you implement the solution you propose to delay loading the Matrix package?
When done I will merge and submit the new version to CRAN.
Again thanks a lot & all the best,
Björn

aitap committed:
Delay loading Matrix to first use

Instead of importing functions from Matrix and paying the cost up front,
load Matrix when txtimage() is called for the first time in the session.

aitap commented:
There you go:

```r
system.time(library(txtplot))
#    user  system elapsed
#   0.004   0.000   0.004

# first call:
system.time(txtimage(matrix(10, 200, 200)))
#    user  system elapsed
#   0.743   0.059   0.803

# subsequent calls:
N <- 20
system.time(for (i in 1:N) txtimage(matrix(10, 200, 200))) / N
#    user  system elapsed 
#  0.0719  0.0150  0.0870 
```
